web/web.go

```diff
@@ -173,40 +172,42 @@ func Run(host string, routes http.Handler, readTimeout, writeTimeout time.Durati
 	}
 
 	// We want to report the listener is closed.
-	var wg sync.WaitGroup
-	wg.Add(1)
+	done := make(chan struct{})
 
 	// Start the listener.
 	var err error
 	go func() {
```

Comment:
@wyattjoh @jcbwlkr @abeltay 

This is a smell to me because on the return we are using this variable.

Question, What if the return called server.ListenAndServe() and we put the OS signal code in a goroutine? Would that clean up the need for this done channel?

Reply:
The problem is that calling server.Shutdown first closes any open listeners. This is in the doc for shutdown

> When Shutdown is called, Serve, ListenAndServe, and ListenAndServeTLS immediately return ErrServerClosed. Make sure the program doesn't exit and waits instead for Shutdown to return.

Reply:
I see, I am sorry I understand.  We need to add this comment to the code.

It is the call to Shutdown() we need to wait for.

Reply:
@abeltay this code is not respecting that comment. How can we change this then?

Reply:
@ardan-bkennedy, may I know which comment above are you referring to?

Reply:
> When Shutdown is called, Serve, ListenAndServe, and ListenAndServeTLS immediately return ErrServerClosed. Make sure the program doesn't exit and waits instead for Shutdown to return.

The code is not respecting this comment.

Reply:
@abeltay Maybe we need to refactor this from scratch. Pretend you were implementing this section of code for the first time. Knowning we need to wait for Shutdown() to complete to return. Knowing the port might be in use. Knowing what you know.

Reply:
Typically, whenever I need to handle the server start/run/shutdown I'd do something like this:

```go
func serve(handler http.Handler) error {

	// Create the errors channel.
	serverErrors := make(chan error, 1)

	// Create the server.
	srv := &http.Server{
		Addr:        ":8080",
		Handler:     handler,
	}

	// Start listening on the server.
	go func() {
		serverErrors <- srv.ListenAndServe()
	}()

	// Listen for an interrupt signal from the OS.
	sig := make(chan os.Signal, 1)
	signal.Notify(sig, os.Interrupt)

	// block until the interrupt signal is recieved
	for {
		select {
		case err := <-serverErrors:
			if err != nil {
				log.Error("main", "CLIServe", err, "Error occurred while trying to serve")
				return err
			}
		case s := <-sig:
			log.User("main", "CLIServe", "Captured %v. Exiting...", s)

			// Shut down gracefully, but wait no longer than 5 seconds before halting.
			ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
			defer cancel()

			if err := srv.Shutdown(ctx); err != nil {
				if err := srv.Close(); err != nil {
					return err
				}
			}

			log.User("main", "CLIServe", "Servers drained and stopped")

			// Complete shutdown was successful.
			return nil
		}
	}
}
```

Reply:
@abeltay this is the logic we need. It using a single channel, it hold the calling goroutine in the right places and it would solve your "port in use" bug.

What do you think?

Reply:
I understand what you mean, it looks better. Thanks for the clarification. I'll give it a shot on this code.